firewall_linux.go

```diff
@@ -3,25 +3,38 @@ package libnetwork
 import (
 	"github.com/docker/libnetwork/iptables"
 	"github.com/sirupsen/logrus"
+	"sync"
 )
 
 const userChain = "DOCKER-USER"
 
-func (c *controller) arrangeUserFilterRule() {
-	c.Lock()
-	arrangeUserFilterRule()
-	c.Unlock()
-	iptables.OnReloaded(func() {
+var (
+	ctrl          *controller = nil
+	userChainOnce sync.Once
```

Comment:
yes, cal_ once is the mean to goal of not setup DOCKER-USER chain multiple times.

Reply:
Right, but effectively we're making controller a singleton(-like) construct, which makes the testing difficult (see above), but also makes the first controller that's created "magic" as that's the controller that now owns iptables.

So, wondering if there should be a `sync.Once` _per controller_, and only call if it iptables is enabled. I see `arrangeUserFilterRule` is already (largely) idempotent (as in: it stops adding the return/jump rules if the chain already existed).

Is `iptables` enabled/disabled a _global_ option, or one _per controller_ (wondering) 🤔

Reply:
This fix assumes there shall be only one instance of controller per docker engine (instantiated moby/moby/daemon/daemon_unix.go). I'd think (please correct me) many things will be broken if multiple libnetwork controllers are created within the same engine.

Reply:
in another word, iptables enabled/disabled is a global option, it is executed based on singleton controller configuration. Hope this make sense.

Reply:
I realise having multiple controllers is not a realistic scenario; it's not what the code reflects though; there's a `New()`, which "creates a new instance of network controller", so it's not a singleton.

If:

- `iptables` enable/disabled is not something that can be changed during the lifetime of a controller (wether that be a single one, or multiple)
- `arrangeUserFilterRule` is idempotent (even _if_ multiple controllers were instantiated, they would not trip over each other, because they'd only mutate iptables if the `DOCKER-USER` chain doesn't exist)


Then, why not setup the DOCKER-USER chain when instantiating the controller? (I don't think a `sync.once` would even be needed in that case)

Reply:
sorry I misunderstood u. I have moved invocation of setupArrangeUserFilterRule to controller.go:New(), and do away with sync.once.
Thx, Su